grp/classic.gi

```diff
@@ -1917,6 +1917,122 @@ InstallMethod( Omega,
     { filt, e, d, R } -> OmegaCons( filt, e, d, Size( R ) ) );
 
 
+#############################################################################
+##
+#F  WallForm( <form>, <m>, <fld> ) . . . . . . . . . . . compute the wall of <m>
+##
+# Input: <m> is an element of the orthogonal group which is given by the form <form>,
+#        <fld> is the finite field of the orthogonal group (in which <m> is defined)
+# Output: The Wall Form of <m> correpsponding to <form>
+# Computes the Wall Form of a matrix. The definition can be found in
+# [Taylor, page 163].
+BindGlobal( "WallForm", function( form, m, fld )
+    local id,  w,  b,  p,  i,  x,  j, d;
+
+    id := One( m );
+
+    # compute a base for Image(id-m) which is a subset of the rows of (id - m)
+    # We also store the index of the rows (corresponding to w) in p
+    w := id - m;
+    b := [];
+    p := [];
+    for i in [ 1 .. Length(w) ]  do
+        if Length(b) = 0  then
+            if w[i] <> 0*w[i]  then
+                Add( b, w[i] );
+                Add( p, i );
+            fi;
+        elif RankMat(b) <> RankMat(Concatenation(b,[w[i]]))  then
+            Add( b, w[i] );
+            Add( p, i );
+        fi;
+    od;
```

Comment:
This computes a lot of ranks. We can improve this slightly by adding a local variable `rank`, which is zero at the start; and then do this:
```suggestion
    p := [];
    rank := 0;
    for i in [ 1 .. Length(w) ]  do
        # add a new row and see if that increases the rank
        Add( b, w[i] );
        if RankMat(b) > rank then
            Add( p, i );
            rank := rank + 1;
        else
            Remove( b ); # rank was not increased, so remove the added row again
        fi;
    od;
```
Of course one can do even better, by also keeping a semi-echelonized version of the matrix `b` around, then sifting the "new" vectors through that; if the result is zero we skip the vector; otherwise we add it to `b` and can add the sifted vector to the semi-echelonized version. Unfortunately GAP seems to lack good primitives for this (something that should be changed); the closes I know of are `CloseMutableBasis` and `IsContainedInSpan`, but they end up reducing the same vector twice. The `cvec` package has a better API for this: `CleanRow( b, v, true, fail )` tests if `v` is in the span of the matrix `b` which must be in semi-echelon form, and return a corresponding boolean, but it also grows `b` if `v` is not yet in there.